airflow/jobs.py

```diff
@@ -2436,8 +2461,17 @@ def _collect_errors(self, ti_status, session=None):
         return err
 
     @provide_session
-    def _execute_for_run_dates(self, run_dates, ti_status, executor, pickle_id,
-                               start_date, session=None):
+    def _execute_for_run_dates(
+        self,
+        run_dates,
+        ti_status,
+        executor,
+        pickle_id,
+        start_date,
+        dagruns_pbar,
```

Comment:
Conceptually the progress bar is a CLI-specific component and the job object itself could be used in a different context where no CLI is involved. A more generic approach would be for the caller to pass some sort of `progress_changed` callback.

I'm thinking about the upcoming REST CLI and how any of this would work. It seems like most of this will have to be stripped out of refactored heavily. Let's make sure that the changes beyond the CLI will work in the REST CLI near future.